File: python/SkyObject.py
#!/usr/bin/python

import Degree
from astro import *

class SkyObject(object):
	def __init__(self):
		#Geocentric Equitorial Coordinates (in degrees)
		self.alpha = None
		self.delta = None
		self.M = None
		self.name = None
		self.distance = None

	def setName(self,Name):
		if type(Name) != str:
			Name = str(Name)
		self.name = Name

	def Name(self):
		return self.name

	def getAltAz(self,Lat,Lng,date):
		H = date.apparentSidereal() - Lng - self.alpha
		H.normalize()
	
		y = sin(H)
		x = cos(H)*sin(Lat) - tan(self.delta)*cos(Lat)
		Az = atan2(y, x)
		Alt = asin(sin(Lat)*sin(self.delta) + cos(Lat)*cos(self.delta)*cos(H))

		return (Alt,Az)

	def setRaDec(self,RA,Dec):
		if type(RA) == float:
			RA = Degree.new(RA)
		if type(Dec) == float:
			Dec = Degree.new(Dec)

		self.alpha = RA
		self.delta = Dec
	
	def getRaDec(self):
		return (self.alpha, self.delta)

	def setMag(self,M):
		self.M = M

	def getMag(self):
		return self.M

	def setDistance(self,dist):
		self.distance = dist

	def getDistance(self):
		return self.distance

File: tsc/src/SkyObject/SkyObject.cpp
/**
 * File: SkyObject.cpp
 * Project: tsc - TinyStarChart
 * Creator: jbangelo
 *
 * Description: Defines an interface for all sky object objects
 *
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 jbangelo
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
#include "SkyObject/SkyObject.h"
#include "Math/AstroMath.h"

using tsc::SkyObject::SkyObject;
using tsc::Math::AstroMath;

SkyObject::SkyObject()
        : _name(""),
          _geocentricEquatorial(0.0, 0.0),
          _M(0.0),
          _dist(0.0)
{

}

SkyObject::SkyObject(string name, real RA, real Dec, real M, real dist)
        : _name(name),
          _geocentricEquatorial(RA, Dec),
          _M(M),
          _dist(dist)
{

}

SkyObject::~SkyObject()
{

}

std::string SkyObject::getName()
{
    return _name;
}

HorizontalCoords SkyObject::getHorizontalCoords(LatLng location, Stardate date)
{
    HorizontalCoords horiz;
    real x, y;

    Degree H = date.apparentSidereal() - location.lng
            - _geocentricEquatorial.ra;
    H.normalize();

    y = Degree::sin(H);
    x = Degree::cos(H) * Degree::sin(location.lat)
            - Degree::tan(_geocentricEquatorial.dec)
                    * Degree::cos(location.lat);
    horiz.az = Degree::atan2(y, x);
    horiz.alt = Degree::asin(
            Degree::sin(location.lat) * Degree::sin(_geocentricEquatorial.dec)
                    + Degree::cos(location.lat)
                            * Degree::cos(_geocentricEquatorial.dec)
                            * Degree::cos(H));

    return horiz;
}

EquatorialCoords SkyObject::getGeocentricEquatorialCoords()
{
    return _geocentricEquatorial;
}

real SkyObject::getMag()
{
    return _M;
}

real SkyObject::getDistance()
{
    return _dist;
}

